Sources/App/Commands/ReconcilePackageListCommand.swift

```diff
@@ -0,0 +1,116 @@
+import Vapor
+import Fluent
+import FluentPostgreSQL
+
+final class ReconcilePackageListCommand: Command
+{
+  static let name = "reconcile_package_list"
+  
+  var arguments: [CommandArgument] = []
+  var options: [CommandOption] = []
+  var help = ["Synchronise the database with the master package list from GitHub."]
+  
+  func run(using context: CommandContext) throws -> Future<Void>
+  {
+    // Grab the GitHub master package list and the database's current package list
+    let masterPackageList = try fetchMasterPackageList(context)
+    let currentPackageList = try fetchCurrentPackageList(context)
+    
+    // Reconcile the package lists and add/remove any new/deleted packages
+    return masterPackageList.and(currentPackageList).flatMap { (master, current) in
+      let reconciler = PackageListReconciler(masterPackageList: master, currentPackageList: current)
+      
+      return context.container.withPooledConnection(to: .psql) { database in
+        // Add any new packages
+        let packageAdditions = reconciler.packagesToAdd.map { url -> Future<Package> in
+          let package = Package(url: url)
+          return package.create(on: database)
+        }
+        
+        // Delete any removed packages
+        let packageDeletions = reconciler.packagesToDelete.map { url in
+          return Package.findByUrl(on: database, url: url).map { package -> Future<Void> in
+            print("Deleting \(package.url)")
+            return package.delete(on: database).map {
+              print("Deleted \(package.url)")
```

Comment:
This `print` is never executed, even though the `print` two lines above it is. I don't know why.

Reply:
That would suggest the delete never actually returns - are you sure they get deleted from the database?

Reply:
Positive. If you check this code out and run it without modification you'll see (something like) this in the log:

```
[ INFO ] Migrating 'psql' database (/Users/dave/Library/Developer/Xcode/DerivedData/Server-bcflpguawrtkznajgdsdxsgtueua/SourcePackages/checkouts/fluent/Sources/Fluent/Migration/MigrationConfig.swift:69)
[ INFO ] Migrations complete (/Users/dave/Library/Developer/Xcode/DerivedData/Server-bcflpguawrtkznajgdsdxsgtueua/SourcePackages/checkouts/fluent/Sources/Fluent/Migration/MigrationConfig.swift:73)
2020-03-18 15:40:54.146579+0000 Run[38757:5176485] SecTaskLoadEntitlements failed error=22 cs_flags=20, pid=38757
2020-03-18 15:40:54.146664+0000 Run[38757:5176485] SecTaskCopyDebugDescription: Run[38757]/0#-1 LF=0
Deleting https://github.com/SwiftNIOExtras/swift-nio-redis.git
Deleting https://github.com/vapor-community/leaf-provider.git
Deleting https://github.com/johnsusek/flowstack.git
Deleting https://github.com/vapor-community/vapormonitoring.git
Deleting https://github.com/vapor/mysql-kit.git
Deleting https://github.com/Kuniwak/MemoryLeakTestKit.git
Deleting https://github.com/vapor/postgresql.git
Program ended with exit code: 0
```

Reply:
Lots of `Deleting`, no `Deleted`.

Reply:
Also, just in case it's not obvious I also checked the database after running this and no records had ever been deleted.

Reply:
Which would explain why the second print is never called. Try adding `databases.enableLogging(on: .psql)` in your configure.swift to see what the actual query being executed is

Reply:
> databases.enableLogging(on: .psql)

That's a useful tip! 🤯

OK, so it looks like it's only executing one delete, which is weird:

```
[psql] [2020-03-18 16:14:14 +0000] SELECT current_setting('server_version') AS "version" []
[ INFO ] Migrating 'psql' database (/Users/dave/Library/Developer/Xcode/DerivedData/Server-bcflpguawrtkznajgdsdxsgtueua/SourcePackages/checkouts/fluent/Sources/Fluent/Migration/MigrationConfig.swift:69)
[psql] [2020-03-18 16:14:15 +0000] SELECT COUNT(*) AS "fluentAggregate" FROM "fluent" []
[psql] [2020-03-18 16:14:15 +0000] SELECT * FROM "fluent" ORDER BY "fluent"."batch" DESC LIMIT 1 OFFSET 0 []
[psql] [2020-03-18 16:14:15 +0000] SELECT * FROM "fluent" WHERE "fluent"."name" = ($1) LIMIT 1 OFFSET 0 ["Package" (TEXT)]
[ INFO ] Migrations complete (/Users/dave/Library/Developer/Xcode/DerivedData/Server-bcflpguawrtkznajgdsdxsgtueua/SourcePackages/checkouts/fluent/Sources/Fluent/Migration/MigrationConfig.swift:73)
[psql] [2020-03-18 16:14:15 +0000] SELECT * FROM "packages" []
2020-03-18 16:14:15.040738+0000 Run[39787:5206950] SecTaskLoadEntitlements failed error=22 cs_flags=20, pid=39787
2020-03-18 16:14:15.040814+0000 Run[39787:5206950] SecTaskCopyDebugDescription: Run[39787]/0#-1 LF=0
[psql] [2020-03-18 16:14:15 +0000] SELECT * FROM "packages" WHERE "packages"."urlString" = ($1) LIMIT 1 OFFSET 0 ["https://github.com/rymcol/swiftcron-example.git" (TEXT)]
[psql] [2020-03-18 16:14:15 +0000] SELECT * FROM "packages" WHERE "packages"."urlString" = ($1) LIMIT 1 OFFSET 0 ["https://github.com/appcron/acinteractor.git" (TEXT)]
Deleting https://github.com/rymcol/swiftcron-example.git
[psql] [2020-03-18 16:14:15 +0000] SELECT * FROM "packages" WHERE "packages"."urlString" = ($1) LIMIT 1 OFFSET 0 ["https://github.com/juri/graphqler.git" (TEXT)]
Deleting https://github.com/appcron/acinteractor.git
[psql] [2020-03-18 16:14:15 +0000] DELETE FROM "packages" WHERE "packages"."id" = ($1) [AC034449-873A-4A08-ADAD-65F4E845137A (UUID)]
Deleting https://github.com/juri/graphqler.git
Program ended with exit code: 0
```

It does three `SELECT` statements there and you can see three `print`s being executed too. Then it does one `DELETE` at the end. The UDID of the `DELETE` matches the *first* package (swiftcron-example.git).

Maybe it's possible that `DELETE` is failing?

Reply:
Just checked and the permissions on the database role look good, so I don't think it's that.

Reply:
It's the same issue as before: you want a `flatMap` instead of a `map`, as you're chaining the `package.delete` Future to the `findByUrl` future.